### ISSUE TITLE
Warnings & Errors in various configurations

### ISSUE THREAD

**Issue by helmesjo** [stephenberry/glaze#1037]
Since I'm maintaining the [`build2` package of `glaze`](https://github.com/build2-packaging/glaze) I just ran it through the (free) CI servers and figured I'll let you know which has warnings & which has errors (in case you'd rather just tackle them in one go).

See build status [here](https://ci.stage.build2.org/@c5c2139e-04ad-48b3-97fd-918e34e4c783?builds=&pv=&th=*&tg=&tc=&pc=&rs=*) for `v2.6.8`.

**Edit.**
Obviously you can ignore the warnings coming from `Eigen` headers.
I should also mention that there appears to be a sporadic error in the `jsonrpc_test` on Windows (unless it's something wrong specifically with my `build2` package).

**Comment by stephenberry** [comment]
Thanks for sharing. Most of these warnings are from Eigen or asio, but I did spot one potential under-read issue for glaze flags, which I'll look into.

Could you point me to the cases of the sporadic error in `jsonrpc_test` on Windows?

Glaze also runs CI on a bunch of platforms and configurations under its [GitHub Actions](https://github.com/stephenberry/glaze/actions). But, your CI tests show that I should probably also be build testing with the `c++26` flag where possible.

**Comment by helmesjo** [comment]
Yep, and especially MSVC complains about a few things. One is the use of `<expected>` with `#if __has_include(<expected>)` which should instead be checked something like this:
```c++
#ifdef __has_include
# if __has_include(<version>)
#   include <version> // bring in feature-macros
# endif
#endif

#if __cpp_expected >= 202202L
#  include <expected>
#endif
```

> Could you point me to the cases of the sporadic error in jsonrpc_test on Windows?

That would be `jsonrpc_test.cpp:43` with `test_name = response:{"jsonrpc":"2.0","result":6,"id":1}`. But I can see that it crashes inside `ut.hpp:1675`:

![image](https://github.com/stephenberry/glaze/assets/2501070/99cc1408-91de-4452-9fe0-b52d15ad1611)
`ss.str()` is `not popping because '@`
```c++

```

**Comment by stephenberry** [comment]
So, I've tried multiple times to fix the MSVC warning for `<expected>`, but older versions of Clang break with these kinds of checks. If you're able to make a pull request that passes, that would be great, but otherwise we've just been living with these warnings and they'll go away in C++23 and beyond.

Yeah, `boost-ut` has some major issues. I've been trying to get them to fix some random [segfaults from simply including the header](https://github.com/boost-ext/ut/issues/618). I've developed a simpler ut library that matches the core API of boost-ut that I will probably switch over to.

**Comment by helmesjo** [comment]
What is the oldest Clang version you support?

**Comment by stephenberry** [comment]
Clang 15 should build. I run actions with 16, 17, & 18

**Comment by stephenberry** [comment]
Unreachable code is annoying to deal with in Glaze because it often means needing to duplicate code, and the compiler will typically simply remove unreachable code. I'm wondering if I should just disable these warnings locally?

**Comment by helmesjo** [comment]
You could probably just be explicit with [`std::unreachable()`](https://en.cppreference.com/w/cpp/utility/unreachable).

**Comment by helmesjo** [comment]
Ok I see. Well, I'll just disable that warning (`/w4702`) for the tests in my `build2` package. Don't think I've seen it in my own project consuming `glaze`.

**Comment by stephenberry** [comment]
Yeah, I saw those. Working on them now.

**Comment by stephenberry** [comment]
To explain a bit more why avoiding unreachable is a challenge:
If Glaze were to use exceptions we could easily make local lambda functions that wrap up behavior and then just call these functions in the appropriate `if constexpr` branches. However, because Glaze does not use exceptions internally any lambda function breaks our ability to return from the function at hand, we just return from the lambda. This requires an additional error check after the lambda is called, to determine if an error occurred in the lambda, instead of just exiting from the function. So, we lose performance by using lambdas.

This is a problem that requires deterministic and non-allocating exception support in C++. There have been proposals to improve exception handling in C++, but I think the broader community needs to see this as more of a critical issue.

**Comment by stephenberry** [comment]
I just merged in #1039, which removes most the warnings across the compilers.

**Comment by stephenberry** [comment]
I'm moving to a new version of [ut](https://github.com/openalgz/ut) to avoid issues in the library, which should solve the problem seen in the jsonrpc tests. In #1041
